cmd/exec_test.go

```diff
@@ -326,13 +367,116 @@ func TestFailFlagCases(t *testing.T) {
 			false
 		}`,
 			},
-			decision: "fail/defined/flag/fail_test",
+			decision:    "fail/defined/flag/fail_test",
+			expectError: false,
 			expected: util.MustUnmarshalJSON([]byte(`{"result": [{
 			"path": "/files/test.json",
 			"result": false
 		}]}`)),
 			fail: true,
 		},
+		{
+			description: "--fail-non-empty with undefined result",
+			files: map[string]string{
+				"files/test.json": `{"foo": 7}`,
+				"bundle/x.rego": `package system
+
+		test_fun := x {
+			x = false
+			x
+		}
+
+		undefined_test {
+			test_fun
+		}`,
+			},
+			expectError: false,
+			expected: util.MustUnmarshalJSON([]byte(`{"result": [{
+			"path": "/files/test.json",
+			"error": {
+				"code": "opa_undefined_error",
+				"message": "/system/main decision was undefined"
+			  }
+		}]}`)),
+			failNonEmpty: true,
+		},
+		{
+			description: "--fail-non-empty with populated result",
+			files: map[string]string{
+				"files/test.json": `{"foo": 7}`,
+				"bundle/x.rego": `package system
+
+		main["hello"]`,
+			},
+			decision:    "",
+			expectError: true,
+			expected: util.MustUnmarshalJSON([]byte(`{"result": [{
+			"path": "/files/test.json",
+			"result": ["hello"]
+		}]}`)),
+			failNonEmpty: true,
+		},
+		{
+			description: "--fail-non-empty with true boolean result",
+			files: map[string]string{
+				"files/test.json": `{"foo": 7}`,
+				"bundle/x.rego": `package fail.defined.flag
+
+               some_function {
+                       input.foo == 7
+               }
+
+               default fail_test := false
+               fail_test {
+                       some_function
+               }`,
+			},
+			decision:    "fail/defined/flag/fail_test",
+			expectError: true,
+			expected: util.MustUnmarshalJSON([]byte(`{"result": [{
+			"path": "/files/test.json",
+			"result": true
+		}]}`)),
+			failNonEmpty: true,
+		},
+		{
+			description: "--fail-non-empty with false boolean result",
+			files: map[string]string{
+				"files/test.json": `{"foo": 7}`,
+				"bundle/x.rego": `package fail.defined.flag
+
+		default fail_test := false
+		fail_test {
+			false
+		}`,
+			},
+			decision:    "fail/defined/flag/fail_test",
+			expectError: true,
+			expected: util.MustUnmarshalJSON([]byte(`{"result": [{
+			"path": "/files/test.json",
+			"result": false
+		}]}`)),
+			failNonEmpty: true,
+		},
+		{
+			description: "--fail-non-empty with an empty array",
+			files: map[string]string{
+				"files/test.json": `{"foo": 7}`,
+				"bundle/x.rego": `package fail.defined.flag
+
+		default fail_test := ["something", "hello"]
+		fail_test := [] if {
+			input.foo == 7
+		}`,
+			},
+			decision:    "fail/defined/flag/fail_test",
+			expectError: false,
+			expected: util.MustUnmarshalJSON([]byte(`{"result": [{
+			"path": "/files/test.json",
+			"result": []
+		}]}`)),
+			failNonEmpty: true,
+		},
```

Comment:
It would be good to add a test case for an empty set coming from a partial rule, i.e. something like:

```rego
package fail.non.empty.flag

fail_test[message] {
    false
    message := "not gonna happen"
}
```